### ISSUE TITLE
Idea: PubSub/MessagingQueues abstraction

### ISSUE THREAD

**Issue by Temikus** [fog/fog#3903]
@selmanj had expressed an idea for a new abstraction of messaging queues to Fog.

Some sort of abstract messaging api in the top-level fog project, so that one could swap out Pubsub and AWS sqs for example, without changing much code.

What do you guys think?

**Comment by geemus** [comment]
Could be interesting. I think getting pubsub and/or sqs well implemented first (I think we have sqs?) first would be important to give us a baseline. Also, having a 3rd one to draw from would make me a little more comfortable (if there is one that makes sense). With just one or two there is a bit more danger of overfitting.

**Comment by geemus** [comment]
Broadly/generally speaking, does sound interesting and like it could be a fit though.

**Comment by selmanj** [comment]
Thanks Artem for filing this.

I admit the idea is somewhat half-baked; there are fairy important differences with how messaging works that I worry it wouldn't be able to build an appropriate abstraction layer on top of them (for example, in Pubsub you can create n subscriptions for a topic, whereas in SQS you may only subscribe to a single topic; I also worry about message sizes, semantics around how long the message remains invisible, dead letter handling, etc...)

**Comment by geemus** [comment]
Yeah, it may be feasible, but still a good deal of stuff to think/work through.
